// pusher-chatkit-server/target/chatkit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var pusher_platform_node_1 = require("pusher-platform-node");
var permissions_1 = require("./permissions");
var utils_1 = require("./utils");
;
;
;
;
var TOKEN_EXPIRY_LEEWAY = 30;
var Chatkit = /** @class */ (function () {
    function Chatkit(options) {
        var instanceLocator = options.instanceLocator, key = options.key, port = options.port, host = options.host, client = options.client;
        var apiInstanceOptions = ({
            locator: instanceLocator,
            key: key,
            port: port,
            host: host,
            client: client,
            serviceName: 'chatkit',
            serviceVersion: 'v1',
        });
        var authorizerInstanceOptions = ({
            locator: instanceLocator,
            key: key,
            port: port,
            host: host,
            client: client,
            serviceName: 'chatkit_authorizer',
            serviceVersion: 'v1',
        });
        this.instanceLocator = instanceLocator;
        this.apiInstance = new pusher_platform_node_1.Instance(apiInstanceOptions);
        this.authorizerInstance = new pusher_platform_node_1.Instance(authorizerInstanceOptions);
    }
    // Token generation
    Chatkit.prototype.authenticate = function (authPayload, userId) {
        return this.apiInstance.authenticate(authPayload, { userId: userId });
    };
    Chatkit.prototype.generateAccessToken = function (authOptions) {
        return this.apiInstance.generateAccessToken(authOptions);
    };
    // User interactions
    Chatkit.prototype.createUser = function (id, name, avatarURL, customData) {
        return this.apiInstance.request({
            method: 'POST',
            path: "/users",
            headers: {
                'Content-Type': 'application/json'
            },
            body: {
                id: id,
                name: name,
                avatar_url: avatarURL,
                custom_data: customData,
            },
            jwt: this.getServerToken(),
        }).then(function () { });
    };
    Chatkit.prototype.deleteUser = function (id) {
        return this.apiInstance.request({
            method: 'DELETE',
            path: "/users/" + id,
            jwt: this.getServerToken(),
        }).then(function () { });
    };
    Chatkit.prototype.getUsers = function () {
        return this.apiInstance.request({
            method: 'GET',
            path: "/users",
            jwt: this.getServerToken(),
        }).then(function (res) {
            return JSON.parse(res.body);
        });
    };
    Chatkit.prototype.getUsersByIds = function (userIds) {
        return this.apiInstance.request({
            method: 'GET',
            path: "/users_by_ids",
            qs: {
                user_ids: userIds.join(','),
            },
            jwt: this.getServerToken(),
        }).then(function (res) {
            return JSON.parse(res.body);
        });
    };
    // Room interactions
    Chatkit.prototype.getRoom = function (roomId) {
        return this.apiInstance.request({
            method: 'GET',
            path: "/rooms/" + roomId,
            jwt: this.getServerToken(),
        }).then(function (res) {
            return JSON.parse(res.body);
        });
    };
    Chatkit.prototype.getRoomMessages = function (roomId, initialId, direction, limit) {
        var qs = {};
        if (initialId) {
            qs['initial_id'] = initialId;
        }
        if (direction) {
            qs['direction'] = direction;
        }
        if (limit) {
            qs['limit'] = limit;
        }
        return this.apiInstance.request({
            method: 'GET',
            path: "/rooms/" + roomId + "/messages",
            jwt: this.getServerToken(),
            qs: qs,
        }).then(function (res) {
            return JSON.parse(res.body);
        });
    };
    Chatkit.prototype.getRooms = function () {
        return this.apiInstance.request({
            method: 'GET',
            path: "/rooms",
            jwt: this.getServerToken(),
        }).then(function (res) {
            return JSON.parse(res.body);
        });
    };
    Chatkit.prototype.createRoom = function (userId, options) {
        var jwt = this.generateAccessToken({
            userId: userId,
            su: true,
        });
        var name = options.name, isPrivate = options.isPrivate, userIds = options.userIds;
        var roomPayload = {
            name: name,
            private: isPrivate || false,
        };
        if (userIds && userIds.length !== 0) {
            roomPayload['user_ids'] = userIds;
        }
        return this.apiInstance.request({
            method: 'POST',
            path: '/rooms',
            jwt: jwt.token,
            body: roomPayload,
        }).then(function (res) {
            return JSON.parse(res.body);
        });
    };
    // Authorizer interactions
    Chatkit.prototype.createRoomRole = function (name, permissions) {
        permissions.forEach(function (perm) {
            if (permissions_1.validRoomPermissions.indexOf(perm) < 0) {
                throw new Error("Permission value \"" + perm + "\" is invalid");
            }
        });
        return this.createRole(name, 'room', permissions);
    };
    Chatkit.prototype.createGlobalRole = function (name, permissions) {
        permissions.forEach(function (perm) {
            if (permissions_1.validGlobalPermissions.indexOf(perm) < 0) {
                throw new Error("Permission value \"" + perm + "\" is invalid");
            }
        });
        return this.createRole(name, 'global', permissions);
    };
    Chatkit.prototype.createRole = function (name, scope, permissions) {
        return this.authorizerInstance.request({
            method: 'POST',
            path: "/roles",
            headers: {
                'Content-Type': 'application/json'
            },
            body: { scope: scope, name: name, permissions: permissions },
            jwt: this.getServerToken(),
        }).then(function () { });
    };
    Chatkit.prototype.deleteGlobalRole = function (roleName) {
        return this.authorizerInstance.request({
            method: 'DELETE',
            path: "/roles/" + roleName + "/scope/global",
            jwt: this.getServerToken(),
        }).then(function () { });
    };
    Chatkit.prototype.deleteRoomRole = function (roleName) {
        return this.authorizerInstance.request({
            method: 'DELETE',
            path: "/roles/" + roleName + "/scope/room",
            jwt: this.getServerToken(),
        }).then(function () { });
    };
    Chatkit.prototype.assignGlobalRoleToUser = function (userId, roleName) {
        return this.assignRoleToUser(userId, roleName);
    };
    Chatkit.prototype.assignRoomRoleToUser = function (userId, roleName, roomId) {
        return this.assignRoleToUser(userId, roleName, roomId);
    };
    Chatkit.prototype.assignRoleToUser = function (userId, roleName, roomId) {
        return this.authorizerInstance.request({
            method: 'PUT',
            path: "/users/" + userId + "/roles",
            headers: {
                'Content-Type': 'application/json'
            },
            body: { name: roleName, room_id: roomId },
            jwt: this.getServerToken(),
        }).then(function () { });
    };
    Chatkit.prototype.getUserRoles = function (userId) {
        return this.authorizerInstance.request({
            method: 'GET',
            path: "/users/" + userId + "/roles",
            jwt: this.getServerToken(),
        }).then(function (res) {
            return JSON.parse(res.body);
        });
    };
    Chatkit.prototype.removeGlobalRoleForUser = function (userId) {
        return this.authorizerInstance.request({
            method: 'DELETE',
            path: "/users/" + userId + "/roles",
            headers: {
                'Content-Type': 'application/json'
            },
            jwt: this.getServerToken(),
        }).then(function () { });
    };
    Chatkit.prototype.removeRoomRoleForUser = function (userId, roomId) {
        return this.authorizerInstance.request({
            method: 'DELETE',
            path: "/users/" + userId + "/roles",
            headers: {
                'Content-Type': 'application/json'
            },
            qs: { room_id: roomId },
            jwt: this.getServerToken(),
        }).then(function () { });
    };
    Chatkit.prototype.getPermissionsForGlobalRole = function (roleName) {
        return this.authorizerInstance.request({
            method: 'GET',
            path: "/roles/" + roleName + "/scope/global/permissions",
            jwt: this.getServerToken(),
        }).then(function (res) {
            return JSON.parse(res.body);
        });
    };
    Chatkit.prototype.getPermissionsForRoomRole = function (roleName) {
        return this.authorizerInstance.request({
            method: 'GET',
            path: "/roles/" + roleName + "/scope/room/permissions",
            jwt: this.getServerToken(),
        }).then(function (res) {
            return JSON.parse(res.body);
        });
    };
    Chatkit.prototype.updatePermissionsForGlobalRole = function (roleName, permissionsToAdd, permissionsToRemove) {
        var permsToCheck = permissionsToAdd.concat(permissionsToRemove);
        permsToCheck.forEach(function (perm) {
            if (permissions_1.validGlobalPermissions.indexOf(perm) < 0) {
                throw new Error("Permission value \"" + perm + "\" is invalid");
            }
        });
        return this.updatePermissionsForRole(roleName, 'global', permissionsToAdd, permissionsToRemove);
    };
    Chatkit.prototype.updatePermissionsForRoomRole = function (roleName, permissionsToAdd, permissionsToRemove) {
        var permsToCheck = permissionsToAdd.concat(permissionsToRemove);
        permsToCheck.forEach(function (perm) {
            if (permissions_1.validRoomPermissions.indexOf(perm) < 0) {
                throw new Error("Permission value \"" + perm + "\" is invalid");
            }
        });
        return this.updatePermissionsForRole(roleName, 'room', permissionsToAdd, permissionsToRemove);
    };
    Chatkit.prototype.getRoles = function () {
        return this.authorizerInstance.request({
            method: 'GET',
            path: "/roles",
            jwt: this.getServerToken(),
        }).then(function (res) {
            return JSON.parse(res.body);
        });
    };
    // General requests
    Chatkit.prototype.apiRequest = function (options) {
        options.jwt = options.jwt || this.getServerToken();
        return this.apiInstance.request(options).then(function (res) {
            return JSON.parse(res.body);
        });
    };
    Chatkit.prototype.authorizerRequest = function (options) {
        options.jwt = options.jwt || this.getServerToken();
        return this.authorizerInstance.request(options).then(function (res) {
            return JSON.parse(res.body);
        });
    };
    Chatkit.prototype.updatePermissionsForRole = function (roleName, scope, permissionsToadd, permissionsToRemove) {
        if (permissionsToadd.length === 0 && permissionsToRemove.length === 0) {
            throw new Error("Either permissionsToAdd or permissionsToRemove is required");
        }
        var body = {};
        if (permissionsToadd.length > 0) {
            body['add_permissions'] = permissionsToadd;
        }
        if (permissionsToRemove.length > 0) {
            body['remove_permissions'] = permissionsToRemove;
        }
        return this.authorizerInstance.request({
            method: 'PUT',
            path: "/roles/" + roleName + "/scope/" + scope + "/permissions",
            headers: {
                'Content-Type': 'application/json'
            },
            body: body,
            jwt: this.getServerToken(),
        }).then(function () { });
    };
    /**
     * This method manages the token for http library and pusher platform
     * communication
     */
    Chatkit.prototype.getServerToken = function () {
        if (this.tokenWithExpiry && this.tokenWithExpiry.expiresAt > utils_1.getCurrentTimeInSeconds()) {
            return this.tokenWithExpiry.token;
        }
        // Otherwise generate new token and its expiration time
        var tokenWithExpiresIn = this.apiInstance.generateAccessToken({ su: true });
        this.tokenWithExpiry = {
            token: tokenWithExpiresIn.token,
            expiresAt: utils_1.getCurrentTimeInSeconds() + tokenWithExpiresIn.expires_in - TOKEN_EXPIRY_LEEWAY,
        };
        return this.tokenWithExpiry.token;
    };
    ;
    return Chatkit;
}());
exports.default = Chatkit;
;
