// pusher-chatkit-server/target/permissions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// All possible permissions
exports.JOIN_ROOM = "room:join";
exports.LEAVE_ROOM = "room:leave";
exports.ADD_ROOM_MEMBER = "room:members:add";
exports.REMOVE_ROOM_MEMBER = "room:members:remove";
exports.CREATE_ROOM = "room:create";
exports.DELETE_ROOM = "room:delete";
exports.UPDATE_ROOM = "room:update";
exports.CREATE_MESSAGE = "message:create";
exports.CREATE_TYPING_EVENT = "room:typing_indicator:create";
exports.SUBSCRIBE_PRESENCE = "presence:subscribe";
exports.UPDATE_USER = "user:update";
exports.GET_ROOM_MESSAGES = "room:messages:get";
exports.GET_USER = "user:get";
exports.GET_ROOM = "room:get";
exports.GET_USER_ROOMS = "user:rooms:get";
exports.CREATE_FILE = "file:create";
exports.GET_FILE = "file:get";
exports.validRoomPermissions = [
    exports.JOIN_ROOM,
    exports.LEAVE_ROOM,
    exports.ADD_ROOM_MEMBER,
    exports.REMOVE_ROOM_MEMBER,
    exports.DELETE_ROOM,
    exports.UPDATE_ROOM,
    exports.CREATE_MESSAGE,
    exports.CREATE_TYPING_EVENT,
    exports.GET_ROOM_MESSAGES,
    exports.CREATE_FILE,
    exports.GET_FILE,
];
exports.validGlobalPermissions = [
    exports.JOIN_ROOM,
    exports.LEAVE_ROOM,
    exports.ADD_ROOM_MEMBER,
    exports.REMOVE_ROOM_MEMBER,
    exports.CREATE_ROOM,
    exports.DELETE_ROOM,
    exports.UPDATE_ROOM,
    exports.CREATE_MESSAGE,
    exports.CREATE_TYPING_EVENT,
    exports.SUBSCRIBE_PRESENCE,
    exports.UPDATE_USER,
    exports.GET_ROOM_MESSAGES,
    exports.GET_USER,
    exports.GET_ROOM,
    exports.GET_USER_ROOMS,
    exports.CREATE_FILE,
    exports.GET_FILE,
];
